### ISSUE TITLE
Create zoom box/magnifying glass

### ISSUE THREAD

**Issue by courtneypattison** [Pattern-Projector/pattern-projector#91]
Create a second pdf viewer at a larger scale that sits on top of the current viewer but is clipped to a smaller size with the transform origin set to the mouse.

**Comment by hluedeke** [comment]
I might try this one - hoping you haven't already starting working on it @courtneypattison!

**Comment by courtneypattison** [comment]
Please do! I haven’t touched it. I think people will be really happy about this one!

**Comment by hluedeke** [comment]
So I learned that pdfjs doesn’t do any sort of optimization or partial rendering. That means that an additional zoom box will always be a full second render. I think we might still be ok here, as these pdfs aren’t typically very full of information, but it’s something to think about. Tiling pdfs won’t fully re-render because we can control which page is rendered based on mouse position.

I’ve started this but it could take me awhile. My nights have been pretty busy lately!

**Comment by courtneypattison** [comment]
No rush! It sucks that pdfjs doesn’t have any optimizations for this case but I have a feeling the performance will be ok. Also, this is making me remember an idea, maybe we should start auto removing margins. The margins on some of these PDFs are wild.

**Comment by hluedeke** [comment]
Are you thinking of using imageData for this? Check the location of the farthest left/top non-background data, and the farthest right/bottom and provide a smaller crop (canvas size) with a predetermined margin?

**Comment by hluedeke** [comment]
I might put a brief pause on this, until the new matrix transformations are in place. I'm thinking those changes are going to affect how I determine coordinates of the zoom box.

**Comment by courtneypattison** [comment]
I think that’s a good idea.

Also, I should probably mention that I’m planning on merging the beta branch into main weekly on Monday/Tuesday to have a more consistent schedule and not be fighting fires all week. Open to suggestions if anyone has other ideas though!

**Comment by hluedeke** [comment]
That's up to you. I will likely only be able to help bug fix on weekends, and even that is going to start getting dicey for me soon. So I probably shouldn't be the reference on how you do your schedule. Whatever works for you!

**Comment by courtneypattison** [comment]
Okie doke, good to know! I was thinking Monday/Tuesday because PP has the most activity on the weekend so I don't want to break things for people on their weekend sewing days lol. I think I'll go with the flow for this release since there's a few things to be completed before it's ready to go.

**Comment by j-huff** [comment]
Once #154 is merged I think the codebase will be in a good place for a feature like this. I changed the 'perspective' matrix so that using transformPoint/s with it will give you results in the same coordinate system as the transformSettings matrix (unit of measure coords), so you won't have to mess with multiplying or dividing by ptDensity as much. The only major change I could forsee would be changing tranformSettings to just be a Matrix, but that wouldn't really mess with the logic too much really.

**Comment by sashasewist** [comment]
For avoiding confusion with how manual calibration works with a user specific calibration zoom, I suggest any zoom type of feature be called "Scale" not Zoom. 
So accurate is "100% scale", then whatever you choose how to do it, typing or presets, the wording should be "120% scale", "90% scale", etc. 
People will love this, and it will be just like printing at a different scale than actual/100%.

**Comment by courtneypattison** [comment]
@sashasewist I'm not sure if this is the right place, but I saw you commented about using Scale instead of Zoom for scaling the PDFs. I love that idea! This feature is a little bit different. It's for making a little magnifying glass that could be used for reading text and looking a lines when stitching overlapping PDFs. I'm going to reopen another issue that was closed for adding zooming to the whole PDF. I think it's time to tackle that one because quite a few questions for PP in the group have to do with scaling nowadays.
